Fortran/UnitTests/descriptor/extends_type_of.f90

```diff
@@ -0,0 +1,100 @@
+module extends_type_of_mod
+
+  type p1
+    integer :: a
+  end type
+
+  type, extends(p1) :: p2
+    integer :: b
+  end type
+ 
+  type k1(a)
+    integer, kind :: a
+  end type
+
+contains
+  subroutine check_extended_type(a, b, expect, label)
+    class(*) :: a
+    class(*) :: b
+    character(len=20) :: label
+    character(len=6) :: check_result
+    logical :: expect
+    if (extends_type_of(a, b) .eqv. expect) then
+      check_result = "passed"
+    else
+      check_result = "FAILED"
+    end if
+    print *, check_result, ": ", trim(label)
+  end subroutine
+  subroutine check_extended_type_pointer(a, b, expect, label)
+    class(*), pointer :: a
+    class(*), pointer :: b
+    character(len=20) :: label
+    character(len=6) :: check_result
+    logical :: expect
+    if (extends_type_of(a, b) .eqv. expect) then
+      check_result = "passed"
+    else
+      check_result = "FAILED"
+    end if
+    print *, check_result, ": ", trim(label)
+  end subroutine
+end module
+
+program test
+  use extends_type_of_mod
+  type(p1), target :: p, r
+  type(p2), target :: q
+  type(k1(10)), target :: k10
+  type(k1(20)), target :: k20
+  integer, target :: i1, i2
+  integer(kind=8), target :: i3
+  real, target :: f
+  
+  class(*), pointer :: null_cp => null()
```

Comment:
If you do similar testing with allocatables, I think you'll find the bug in the implementation.

Reply:
Actually, mirroring the tests to allocatables didn't seem to turn anything up... Which case are you expecting to fail?

Reply:
Maybe I recorded the wrong answer for a test.

Reply:
Your original implementation used `mold` instead of `a` in a Boolean expression.